src/modules/rlm_unpack/rlm_unpack.c

```diff
@@ -59,154 +64,93 @@ fr_dict_attr_autoload_t rlm_unpack_dict_attr[] = {
+				 fr_value_box_list_t *in)
 {
-	bool tainted = false;
-	char *data_name, *data_size, *data_type;
-	char *p;
-	size_t len, input_len, offset;
-	ssize_t slen;
-	fr_type_t type;
-	fr_dict_attr_t const *da;
-	fr_pair_t *vp;;
-	uint8_t const *input;
-	char buffer[256];
-	uint8_t blob[256];
-
-	/*
-	 *	FIXME: copy only the fields here, as we parse them.
-	 */
-	strlcpy(buffer, fmt, sizeof(buffer));
-
-	p = buffer;
-	fr_skip_whitespace(p); /* skip leading spaces */
-
-	data_name = p;
-
-	fr_skip_not_whitespace(p);
-
-	if (!*p) {
-	error:
-		REDEBUG("Format string should be '<data> <offset> <type>' e.g. '&Class 1 integer'");
-	nothing:
-		return -1;
+	size_t		len, input_len, offset;
+	fr_type_t	type;
+	uint8_t const	*input;
+	uint8_t		blob[256];
+	fr_value_box_t	*data_vb = fr_dlist_head(in);
+	fr_value_box_t	*offset_vb = fr_dlist_next(in, data_vb);
+	fr_value_box_t	*type_vb = fr_dlist_next(in, offset_vb);
+	fr_value_box_t	*vb;
+
+	if ((data_vb->type != FR_TYPE_OCTETS) && (data_vb->type != FR_TYPE_STRING)) {
+		REDEBUG("unpack requires the input attribute to be 'string' or 'octets'");
+		return XLAT_ACTION_FAIL;
 	}
 
-	fr_zero_whitespace(p);
-	if (!*p) GOTO_ERROR;
-
-	data_size = p;
-
-	fr_skip_not_whitespace(p);
-	if (!*p) GOTO_ERROR;
-
-	fr_zero_whitespace(p);
-	if (!*p) GOTO_ERROR;
-
-	data_type = p;
-
-	fr_skip_not_whitespace(p);
-	if (*p) GOTO_ERROR;	/* anything after the type is an error */
-
-	/*
-	 *	Attribute reference
-	 */
-	if (*data_name == '&') {
-		fr_pair_t *from_vp;
-
-		if (xlat_fmt_get_vp(&from_vp, request, data_name) < 0) goto nothing;
-
-		if ((from_vp->vp_type != FR_TYPE_OCTETS) &&
-		    (from_vp->vp_type != FR_TYPE_STRING)) {
-			REDEBUG("unpack requires the input attribute to be 'string' or 'octets'");
-			goto nothing;
-		}
-		input = from_vp->vp_octets;
-		input_len = from_vp->vp_length;
-		tainted = from_vp->vp_tainted;
-
-	} else if ((data_name[0] == '0') && (data_name[1] == 'x')) {
+	if ((data_vb->type == FR_TYPE_STRING) && (data_vb->vb_length > 1) &&
+	    (data_vb->vb_strvalue[0] == '0') && (data_vb->vb_strvalue[1] == 'x')) {
 		/*
 		 *	Hex data.
 		 */
-		len = strlen(data_name + 2);
+		len = strlen(data_vb->vb_strvalue + 2);
 		if (len > 0) {
 			fr_sbuff_parse_error_t err;
 
 			input = blob;
 			input_len = fr_hex2bin(&err, &FR_DBUFF_TMP(blob, sizeof(blob)),
-					       &FR_SBUFF_IN(data_name + 2, len), true);
+					       &FR_SBUFF_IN(data_vb->vb_strvalue + 2, len), true);
 			if (err) {
-				REDEBUG("Invalid hex string in '%s'", data_name);
-				goto nothing;
+				REDEBUG("Invalid hex string in '%s'", data_vb->vb_strvalue);
+				return XLAT_ACTION_FAIL;
 			}
 		} else {
-			GOTO_ERROR;
+			REDEBUG("Zero length hex string in '%s'", data_vb->vb_strvalue);
+			return XLAT_ACTION_FAIL;
 		}
+	} else if (data_vb->type == FR_TYPE_STRING) {
+		input = (uint8_t const *)data_vb->vb_strvalue;
+		input_len = data_vb->length;
 	} else {
-		GOTO_ERROR;
+		input = data_vb->vb_octets;
```

Comment:
this is removing the requirement above that `unpack requires the input attribute to be 'string' or 'octets'"`

i.e. it should be

```
} else if (data_vb->type == FR_TYPE_OCTETS) {
   ... do octets stuff
} else {
  complain and return an error
}

Reply:
If the data isn't "octets" form, it might be worth just casting it to octets?  That would allow for things like unpacking 24-bit vendor IDs from ethernet addresses.

I suppose the user could manually add a cast to the xlat, but it might be nice to just do it for them.

Reply:
At the top of the routine there is a test for whether the data is octets or string - anything else throws an error and returns.

I could add a cast for other data types.

Reply:
OK, for now let's just leave it.